Bug 1211932 - Fix problem with escaping for URI

hfiguiere opened the pull request.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b57363dfed6421f544021df3a48f25174cc65a1d

hfiguiere committed:
Bug 1211932 - Generate lighter-weight model object and properly encode URL.

Fetch on non webworker no longer need encodeURI
getSectionName callback are lighter too

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=53e931c1f2f6f87acfb57d38e955569d81a79ef6

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6d55068ed5e8d5c9f8343c7ec49f4ce1a313896c

justindarc commented on apps/music/views/albums/view.js:
```diff
@@ -77,6 +77,7 @@ AlbumsView.prototype.getAlbums = function() {
         .then(response => response.json())
         .then((albums) => {
           albums.forEach((album) => {
```
Let's do a `map()` call here instead to produce a new array and just copy the properties we need to new objects. Jim suggested we do this for search and it seems to work ok there. His reasoning was that the original array and heavy metadata objects would get GC'd and it may end up using less memory.

justindarc replied:
Also, instead of copying the full metadata object over to the new one, just grab `album` and `artist`. This may end up being somewhat of a mini-megapatch, but it'll be worth it.

hfiguiere replied:
Good point. Doing this.

justindarc commented on apps/music/views/artists/view.js:
```diff
@@ -77,6 +77,7 @@ ArtistsView.prototype.getArtists = function() {
         .then(response => response.json())
         .then((artists) => {
           artists.forEach((artist) => {
+            artist.url = '/artist-detail?id=' + encodeURIComponent(artist.name);
```
Ditto here about using `map()` instead.

justindarc commented on apps/music/views/home/view.js:
```diff
@@ -113,7 +115,8 @@ HomeView.prototype.search = function(query) {
   ).then(([unknownTitle, unknownArtist, unknownAlbum]) => {
-    var albumResults = this.fetch('/api/search/album/' + query)
+    var albumResults = this.fetch('/api/search/album/' +
+                                  encodeURIComponent(query))
```
Let's just do `query = encodeURIComponent(query);` at the top of this function so we're not calling it three separate times.

hfiguiere replied:
done

justindarc commented on apps/music/views/player/view.js:
```diff
@@ -119,7 +119,7 @@ PlayerView.prototype.next = function() {
 PlayerView.prototype.share = function() {
   this.getPlaybackStatus().then((status) => {
-    this.fetch('/api/activities/share/' + status.filePath);
+    this.fetch('/api/activities/share/' + encodeURIComponent(status.filePath));
```
I just had a thought about this. So, `this.fetch` comes from the "js/view.js" base `View` class. It either maps the call to a bridge.js method or uses `window.fetch` if Service Workers are enabled. We can probably just put the `encodeURIComponent()` call inside of the `fetch` method in the base `View` class instead of having to do this everywhere.

hfiguiere replied:
Or just remove the ```decodeURIComponent()``` in fetch() altogether with the encode

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=44b2a53545701082bfe683e5b7c100820ac5dc4c

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=01c39c330887f7358a092f9b83a35edb7af60e11

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4e7fe96a8c96e14143f156f046bb6d0f52e6a9c2

wilsonpage commented on apps/music/views/albums/view.js:
```diff
@@ -76,14 +76,16 @@ AlbumsView.prototype.getAlbums = function() {
+              album: album.metadata.album || unknownAlbum,
+              unknownAlbum: !album.metadata.album ? true : false,
+              artist: album.metadata.artist
+            };
```
I think a comment explaining what/why this function is doing what it's doing would be nice :)

hfiguiere replied:
ok

justindarc commented on apps/music/views/albums/view.js:
```diff
@@ -36,7 +36,7 @@ var AlbumsView = View.extend(function AlbumsView() {
         }
 
-        var sectionName = item.metadata.album[0].toLowerCase();
+        var sectionName = item.album[0].toLowerCase();
```
Couldn't we just move all of this logic into the `map()` function? Just add a property to the objects called `sectionName`.

hfiguiere replied:
indeed. we could do that.

hfiguiere replied:
bonus point: we don't call l10n in there anymore...

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=11b1abfd6187814aea158733a000d42f86c19319

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ccb9ef72a4423ca96d50cb6d10c7f32534121840

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=259f5ac40c1fa7143a728dd0ddf2041f789e4589

justindarc commented on apps/music/views/songs/view.js:
```diff
@@ -73,11 +71,26 @@ SongsView.prototype.render = function() {
+        .then((songs) => {
+          var items = songs.map((song) => {
+            var title = song.metadata.title;
+            var firstChar = title ? title[0].toLowerCase() : unknown;
```
`var sectionName` to be consistent with other views

hfiguiere replied:
true. cut & paste cargo cult

justindarc commented on apps/music/views/album-detail/index.html:
```diff
@@ -29,9 +29,9 @@
   <gaia-fast-list id="list">
     <template>
-      <a href="/player?id=${name}" data-file-path="${name}">
+      <a href="${url}" data-file-path="${name}">
```
I *think* we can hard-code simply `href="/player"` here and we don't need the `url` property in this case.

justindarc commented on apps/music/views/album-detail/view.js:
```diff
@@ -52,16 +52,20 @@ AlbumDetailView.prototype.getAlbum = function() {
+                unpaddedIndex.format(song.metadata.tracknum) : ''),
+          title: song.metadata.title,
+          name: song.name,
+          url: '/player?id=' + encodeURIComponent(song.name)
```
You can remove this `url` property completely.

justindarc commented on apps/music/views/artist-detail/index.html:
```diff
@@ -29,9 +29,9 @@
   <gaia-fast-list id="list">
     <template>
-      <a href="/player?id=${name}" data-file-path="${name}">
+      <a href="${url}" data-file-path="${name}">
```
Hardcode `href="/player"`

justindarc commented on apps/music/views/artist-detail/view.js:
```diff
@@ -43,15 +43,18 @@ ArtistDetailView.prototype.render = function() {
+            unpaddedIndex.format(song.metadata.tracknum) : '',
+          name: song.name,
+          title: song.metadata.title,
+          url: '/player?id=' + encodeURIComponent(song.name)
```
Remove `url` property completely here

justindarc commented on apps/music/views/home/view.js:
```diff
@@ -61,9 +61,10 @@ HomeView.prototype.render = function() {
     var html = [];
 
     this.albums.forEach((album) => {
+      var url = '/player?id=' + encodeURIComponent(album.name);
```
Remove this var.

justindarc commented on apps/music/views/home/view.js:
```diff
@@ -61,9 +61,10 @@ HomeView.prototype.render = function() {
       var template =
 Sanitizer.createSafeHTML `<a class="tile"
-    href="/player?id=${album.name}"
+    href="${url}"
```
Just hardcode `href="/player"` here

justindarc commented on apps/music/views/home/view.js:
```diff
@@ -160,7 +161,7 @@ HomeView.prototype.search = function(query) {
             subtitle: song.metadata.artist || unknownArtist,
             section:  'songs',
-            url:      '/player?id=' + song.name
+            url:      '/player?id=' + encodeURIComponent(song.name)
```
just hardcode `url: '/player'` here

justindarc commented on apps/music/views/playlist-detail/index.html:
```diff
@@ -31,10 +31,10 @@
     <template>
       <li>
-        <a href="/player?id=${name}" data-file-path="${name}">
+        <a href="${url}" data-file-path="${name}">
```
Hardcode `href="/player"` here

justindarc commented on apps/music/views/playlist-detail/view.js:
```diff
@@ -43,11 +43,15 @@ PlaylistDetailView.prototype.getPlaylist = function() {
+      return songs.map((song, index) => {
+        return {
+          index: unpaddedIndex.format(index + 1),
+          url: '/player?id=' + encodeURIComponent(song.name),
```
You can remove this `url` property completely.

justindarc commented on apps/music/views/songs/index.html:
```diff
@@ -32,10 +32,10 @@
     <a picker-item href="#search" data-icon="search" data-l10n-id="search"></a>
     <template>
-      <a href="/player?id=${name}" data-file-path="${name}">
+      <a href="${url}" data-file-path="${name}">
```
Hardcode `href="/player"` here

justindarc commented on apps/music/views/songs/view.js:
```diff
@@ -73,11 +71,26 @@ SongsView.prototype.render = function() {
+              title: title,
+              artist: song.metadata.artist,
+              name: song.name,
+              url: '/player?id=' + encodeURIComponent(song.name)
```
Remove this `url` property completely.

justindarc commented on apps/music/views/songs/view.js:
```diff
@@ -103,7 +116,7 @@ SongsView.prototype.search = function(query) {
             subtitle: song.metadata.artist || unknownArtist,
             section:  'songs',
-            url:      '/player?id=' + song.name
+            url:      '/player?id=' + encodeURIComponent(song.name)
```
Hardcode `url: '/player'`

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=65ef177a53563fffd5f458ff7eb4339b0e4c3d70

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4fcb37cec925ede1fe2aa8fe97fe4d40df700f7d

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=74cfa3e433069002fbf119f921ae3438f1ab6e7d
